File: test/GSS.Authentication.CAS.E2E.Tests/PageObjects/KeycloakLoginPage.cs
namespace GSS.Authentication.CAS.E2E.Tests.PageObjects;

public class KeycloakLoginPage
{
    private readonly IPage _page;

    public KeycloakLoginPage(IPage page)
    {
        _page = page;
    }

    public ILocator UsernameInput => _page.Locator("#username");
    public ILocator PasswordInput => _page.Locator("#password");
    public ILocator LoginButton => _page.Locator("#kc-login");

    public async Task<bool> IsOnKeycloakLoginPageAsync()
    {
        try
        {
            await UsernameInput.WaitForAsync(new() { Timeout = 5000 });
            return true;
        }
        catch
        {
            return false;
        }
    }

    public async Task LoginAsync(string username, string password)
    {
        await UsernameInput.FillAsync(username);
        await PasswordInput.FillAsync(password);
        await LoginButton.ClickAsync();
    }

    public async Task WaitForRedirectAsync(string expectedUrlPrefix)
    {
        await _page.WaitForURLAsync(url => url.StartsWith(expectedUrlPrefix), new() { Timeout = 10000 });
    }
}

File: e2e/support/keycloak-login-page.ts
import type { Locator, Page } from '@playwright/test';

export class KeycloakLoginPage {
  readonly usernameInput: Locator;
  readonly passwordInput: Locator;
  readonly loginButton: Locator;

  constructor(private readonly page: Page) {
    this.usernameInput = page.locator('#username');
    this.passwordInput = page.locator('#password');
    this.loginButton = page.locator('#kc-login');
  }

  async isOnLoginPage(timeout = 5000): Promise<boolean> {
    return this.usernameInput
      .waitFor({ timeout })
      .then(() => true)
      .catch(() => false);
  }

  async login(username: string, password: string): Promise<void> {
    await this.usernameInput.fill(username);
    await this.passwordInput.fill(password);
    await this.loginButton.click();
  }

  /** Waits until the CAS redirect lands back on the sample app's host, regardless of port (the React sample's dev proxy can hop ports). */
  async waitForRedirect(expectedBaseUrl: string, timeout = 30000): Promise<void> {
    const host = new URL(expectedBaseUrl).hostname;
    await this.page.waitForURL((url) => url.hostname === host, { timeout });
  }
}
